dex/src/main/scala/com/wavesplatform/dex/tool/WaitOffsetTool.scala

```diff
@@ -0,0 +1,76 @@
+package com.wavesplatform.dex.tool
+
+import akka.actor.Scheduler
+import com.wavesplatform.dex.domain.utils.ScorexLogging
+import com.wavesplatform.dex.queue.ValidatedCommandWithMeta.Offset
+import com.wavesplatform.dex.settings.MatcherSettings
+
+import java.util.concurrent.TimeoutException
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.concurrent.duration.Deadline
+import scala.util.{Failure, Success}
+import scala.concurrent.duration._
+
+trait WaitOffsetTool extends ScorexLogging {
+
+  case class TimeCheckTag(currentOffset: Offset, time: Long)
+
+  def waitOffsetReached(
+    getLastOffset: => Future[Offset],
+    getLastProcessedOffset: => Offset,
+    startingLastOffset: Offset,
+    deadline: Deadline,
+    settings: MatcherSettings,
+    scheduler: Scheduler
+  )(implicit ex: ExecutionContext): Future[Unit] = {
+    val interval = settings.waitingQueue.checkInterval
+    val maxWaitingTime = settings.waitingQueue.maxWaitingTime
+
+    def canProcessNewCommands(lastOffset: Offset, currentOffset: Offset, commandsPerSecond: Option[Double]): Boolean =
+      commandsPerSecond.fold(currentOffset >= lastOffset) { k =>
```

Comment:
if i understood right currentOffset >= lastOffset this condition executes only the very first time. Can we do this check before all and then work only with (lastOffset - currentOffset) / k <= maxWaitingTime.toSeconds. It will simplify the whole logic.